### ISSUE TITLE
Add build target for windows

### ISSUE THREAD

**Issue by JJRcop** [cspiegel/audacious-openmpt#2]
I don't know how to get this working with mingw-w64 myself, not very competent with it. I tried just replacing g++ with i686-w64-mingw32-g++-win32, then when that didn't work calling it manually with the same flags, but I honestly don't know what I'm doing, and I'd like to be able to build a DLL out of this.

If you don't use windows then I don't blame you for not bothering with this, it might be too much work if you wouldn't personally use it.

**Comment by cspiegel** [comment]
Building for Windows is a bit more complex than just replacing the compiler, but you may be in luck: I've been experimenting with a CMake-based build system for this plugin, and for another project, I've gotten MinGW builds working with CMake, so I was able to put something together here which at least builds a DLL for me.  I don't have a Windows system handy to test on, but it's worth a try.

I've created a branch called `cmake-mingw` which includes a brief `README.MINGW`.  It's not a trivial process but it's not particularly onerous, either.  If you run into problems, let me know, and I can work on tracking them down.  Once I switch to CMake I'll be happy to keep the MinGW parts in, if they do indeed work.

**Comment by JJRcop** [comment]
I followed the instructions and didn't end up with cas-mpt.dll, no errors though, cmake completed.

Here's all the new files that were created (and also shows what I did):
```
	new file:   aud/include/libaudcore/audstrings.h
	new file:   aud/include/libaudcore/drct.h
	new file:   aud/include/libaudcore/equalizer.h
	new file:   aud/include/libaudcore/export.h
	new file:   aud/include/libaudcore/hook.h
	new file:   aud/include/libaudcore/i18n.h
	new file:   aud/include/libaudcore/i18n.h.orig
	new file:   aud/include/libaudcore/index.h
	new file:   aud/include/libaudcore/inifile.h
	new file:   aud/include/libaudcore/interface.h
	new file:   aud/include/libaudcore/list.h
	new file:   aud/include/libaudcore/mainloop.h
	new file:   aud/include/libaudcore/multihash.h
	new file:   aud/include/libaudcore/objects.h
	new file:   aud/include/libaudcore/playlist.h
	new file:   aud/include/libaudcore/plugin.h
	new file:   aud/include/libaudcore/plugins.h
	new file:   aud/include/libaudcore/preferences.h
	new file:   aud/include/libaudcore/probe.h
	new file:   aud/include/libaudcore/ringbuf.h
	new file:   aud/include/libaudcore/runtime.h
	new file:   aud/include/libaudcore/templates.h
	new file:   aud/include/libaudcore/tinylock.h
	new file:   aud/include/libaudcore/tuple.h
	new file:   aud/include/libaudcore/vfs.h
	new file:   aud/include/libaudcore/vfs_async.h
	new file:   aud/include/libaudcore/visualizer.h
	new file:   aud/lib/libaudcore.dll
	new file:   aud/lib/libopenmpt.dll
	new file:   aud/libopenmpt/libopenmpt.h
	new file:   aud/libopenmpt/libopenmpt.hpp
	new file:   aud/libopenmpt/libopenmpt_config.h
	new file:   aud/libopenmpt/libopenmpt_ext.h
	new file:   aud/libopenmpt/libopenmpt_ext.hpp
	new file:   aud/libopenmpt/libopenmpt_stream_callbacks_buffer.h
	new file:   aud/libopenmpt/libopenmpt_stream_callbacks_fd.h
	new file:   aud/libopenmpt/libopenmpt_stream_callbacks_file.h
	new file:   aud/libopenmpt/libopenmpt_version.h
	new file:   audacious-3.9a-win32.zip
	new file:   build/CMakeCache.txt
	new file:   build/CMakeFiles/3.9.5/CMakeCCompiler.cmake
	new file:   build/CMakeFiles/3.9.5/CMakeCXXCompiler.cmake
	new file:   build/CMakeFiles/3.9.5/CMakeDetermineCompilerABI_C.bin
	new file:   build/CMakeFiles/3.9.5/CMakeDetermineCompilerABI_CXX.bin
	new file:   build/CMakeFiles/3.9.5/CMakeRCCompiler.cmake
	new file:   build/CMakeFiles/3.9.5/CMakeSystem.cmake
	new file:   build/CMakeFiles/3.9.5/CompilerIdC/CMakeCCompilerId.c
	new file:   build/CMakeFiles/3.9.5/CompilerIdC/a.exe
	new file:   build/CMakeFiles/3.9.5/CompilerIdCXX/CMakeCXXCompilerId.cpp
	new file:   build/CMakeFiles/3.9.5/CompilerIdCXX/a.exe
	new file:   build/CMakeFiles/CMakeDirectoryInformation.cmake
	new file:   build/CMakeFiles/CMakeOutput.log
	new file:   build/CMakeFiles/Makefile.cmake
	new file:   build/CMakeFiles/Makefile2
	new file:   build/CMakeFiles/TargetDirectories.txt
	new file:   build/CMakeFiles/cas-mpt.dir/DependInfo.cmake
	new file:   build/CMakeFiles/cas-mpt.dir/build.make
	new file:   build/CMakeFiles/cas-mpt.dir/cmake_clean.cmake
	new file:   build/CMakeFiles/cas-mpt.dir/depend.make
	new file:   build/CMakeFiles/cas-mpt.dir/flags.make
	new file:   build/CMakeFiles/cas-mpt.dir/includes_CXX.rsp
	new file:   build/CMakeFiles/cas-mpt.dir/link.txt
	new file:   build/CMakeFiles/cas-mpt.dir/linklibs.rsp
	new file:   build/CMakeFiles/cas-mpt.dir/objects1.rsp
	new file:   build/CMakeFiles/cas-mpt.dir/progress.make
	new file:   build/CMakeFiles/cmake.check_cache
	new file:   build/CMakeFiles/feature_tests.bin
	new file:   build/CMakeFiles/feature_tests.c
	new file:   build/CMakeFiles/feature_tests.cxx
	new file:   build/CMakeFiles/progress.marks
	new file:   build/Makefile
	new file:   build/cmake_install.cmake
	new file:   libopenmpt-0.3.6+release.dev.winold.vs2017.zip
```

**Comment by cspiegel** [comment]
Oops! I forgot one crucial step. In the build directory, run `make`. That will produce the DLL.

**Comment by JJRcop** [comment]
Okay, let's see how this runs! Sorry for my ignorance by the way.

**Comment by JJRcop** [comment]
Serves me right for having a 64 bit version of audacious' dev files, make note of that

**Comment by JJRcop** [comment]
https://github.com/cspiegel/audacious-openmpt/blob/41e09b627fb88c180bee94b05486c043052857c7/README.MINGW#L45
This is incorrect, I had a build error which was fixed when I moved aud/libopenmpt into aud/include/libopenmpt

I now have a cas-openmpt.dll file and i'm going to boot up windows and test it

**Comment by JJRcop** [comment]
Sorry I haven't responded in a bit, the dll didn't work and I'm not really sure what kind of info i can give you besides this: `ERROR plugin-load.cc:72 [plugin_load]: C:\Program Files (x86)\Audacious\lib\audacious\Input\cas-mpt.dll could not be loaded: 'C:\Program Files (x86)\Audacious\lib\audacious\Input\cas-mpt.dll': %1 is not a valid Win32 application.`

I may need to go to the audacious forums or find another kind of plugin that got win32 right and see what they did, so I can tell you.

**Comment by elsemieni** [comment]
I was able to make a build for windows recently. I used MSYS2 with mingw-w64-i686, and built the two main dependencies from source (libopenmpt and audacious ... this last one requires Qt so the size of the dependencies is quite large, but it's just necessary for the headers and for libaudcore so you can leave some dependencies out). After make install it, it generates an .so file in \mingw32\lib\audacious\Input, depending from libaudcore.dll, some mingw dll (already included in audacious bin folder). 

You must rename it to dll (as makefile is hardcoded to name it with .so extension). That's the plugin you must copy to your lib/audacious/Input directory. Then download libopenmpt binaries for windows x86 from https://lib.openmpt.org/libopenmpt/download/ and place it in bin folder. 


_(For the people who didn't manage well with building systems and such I will attach my compiled audacious-openmpt binary version here for Audacious 4.0.4 windows : [cas-mpt.zip](https://github.com/cspiegel/audacious-openmpt/files/4835122/cas-mpt.zip) Just place it in audacious /lib/audacious/Input folder, and then download https://lib.openmpt.org/files/libopenmpt/dev/libopenmpt-0.5.0+release.dev.win.vs2019.zip and place bin/x86 folder contents to audacious bin folder, rename libopenmpt.dll to libopenmpt-0.dll :) ... Can't guarantee it will work in future versions, but it's better than nothing. )_

**Comment by cspiegel** [comment]
This plugin has been incorporated into the official Audacious plugins pack; I've verified that the Windows release of 4.1 includes the plugin and that it works.  I'll close this issue now since it's no longer necessary to be able to build this manually with MinGW.

I've pushed a new (untested) version of the `cmake-mingw` branch anyway, though, in case somebody wants to try to get that working for some reason.
